File: 23-1/boj/18405.py
import sys
from collections import deque
from copy import deepcopy
input = sys.stdin.readline

# n: 그래프 row, k: 바이러스 종류 갯수
n,k=map(int, input().split())
graph = [list(map(int, input().split())) for _ in range(n)]
s,x,y = map(int, input().split())

dr = [-1,1,0,0]
dc = [0,0,-1,1]

virus = []
que = deque()
t_que = deque()

for i in range(n):
    for j in range(n):
        if graph[i][j]!=0:
            virus.append([i,j, graph[i][j]])
            
virus.sort(key = lambda x : x[2])

for r,c,v in virus:
    que.append([r,c])
            
def bfs(s, que):
    count=1
    while que:
        if count>s:
                return
        r,c = que.popleft()
        for idx in range(4):
            nr, nc = r + dr[idx], c+ dc[idx]
            if not(0<=nr<n and 0<=nc<n):
                continue
            if graph[nr][nc]==0:
                graph[nr][nc] = graph[r][c]
                t_que.append([nr, nc])
        #print("##",count,"##")
        #for i in graph:
        #    print(i)       
        #print("####")
        if len(que)==0:
            count+=1
            que = deepcopy(t_que)
            t_que.clear()
            
            
bfs(s, que)
print(graph[x-1][y-1])

File: 24-study/DFS:BFS/18405.js


const fs = require("fs");
const filePath = process.platform === "linux" ? "/dev/stdin" : "./input.txt";
let lines = fs.readFileSync(filePath).toString().trim().split("\n")


let graph = [];
let virus = [];
let que = [];


// Process the first line to get n and k
const [n, k] = lines[0].split(" ").map(Number);
graph = lines
  .slice(1, 1 + n)
  .map((line) => line.split(" ").map(Number));


// Process the second line to get s, x, and y
const [s, x, y] = lines[lines.length-1].split(" ").map(Number);

// Process the virus data
for (let r = 0; r < graph.length; r++) {
  for (let c = 0; c < graph[0].length; c++) {
    if (graph[r][c]) {
      virus.push([graph[r][c], r, c]);
    }
  }
}
virus.sort((a, b) => a[0] - b[0]);

// Initialize the queue with virus data
  for (let v of virus) {
    que.push([v[1], v[2], 0]);
  }

// The rest of the code remains the same
const dr = [-1, 1, 0, 0];
const dc = [0, 0, -1, 1];

  function bfs() {
    while (que.length) {
      let [r, c, time] = que.shift();

      if (time === s) {
        continue;
      }

      for (let idx = 0; idx < 4; idx++) {
        const nr = r + dr[idx];
        const nc = c + dc[idx];
        if (!(0<=nc && nc<n)||(!(0<=nr && nr<n))) {
          continue;
        }
        if (!graph[nr][nc]) {
          graph[nr][nc] = graph[r][c];
          que.push([nr, nc, time + 1]);
        }
      }
    }
  }

bfs();
console.log(graph[x - 1][y - 1]);